benchmark/modules/src/BM_ocv_vision_factory_test.cpp

```diff
@@ -0,0 +1,83 @@
+// Copyright 2023 Klepsydra Technologies AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <benchmark/benchmark.h>
+#include <spdlog/spdlog.h>
+#include <opencv2/opencv.hpp>
+#include <image_event_stamped.h>
+#include <klepsydra/mem_core/basic_middleware_provider.h>
+#include <klepsydra/vision_ocv/image_data_factory.h>
+
+
+
+static void BM_FactoryPerformanceTest(benchmark::State &state)
+{
+    // vars to test
+    bool withFactory = state.range(0);
+
+    kpsr::vision_ocv::ImageDataFactory imageDataFactory(500, 500, cv::IMREAD_UNCHANGED, "raw_image");
+    auto initF = [&imageDataFactory](kpsr::vision_ocv::ImageDataStamped & e) {
+        e.stamp = std::chrono::high_resolution_clock::now();
+        imageDataFactory.initializerFunction(e.im);
+    };
+    auto cloneF = [&imageDataFactory] (const kpsr::vision_ocv::ImageDataStamped &source,
+                                                     kpsr::vision_ocv::ImageDataStamped &target) {
+        target.stamp = source.stamp;
+        imageDataFactory.eventClonerFunction(source.im, target.im);
+    };
+    double total_duration = 0;
+    std::unique_ptr<kpsr::mem::BasicMiddlewareProvider<kpsr::vision_ocv::ImageDataStamped>> mp;
+    if (withFactory) {
+        mp.reset(new kpsr::mem::BasicMiddlewareProvider<kpsr::vision_ocv::ImageDataStamped>(nullptr,
+          "img", 4, 6, initF, cloneF, false));
+    } else {
+        mp.reset(new kpsr::mem::BasicMiddlewareProvider<kpsr::vision_ocv::ImageDataStamped>(nullptr,
+          "img", 4, 6, nullptr, nullptr, false));
+    }
+    mp->start();
+    auto receiveImGetDuration = [&total_duration](const kpsr::vision_ocv::ImageDataStamped & im) {
+        auto now = std::chrono::high_resolution_clock::now();
+        auto duration = std::chrono::duration_cast<std::chrono::duration<double>>(now - im.stamp).count();
+        total_duration += duration; // maybe should be atomic
```

Comment:
I introduced the changes that you mentioned and compared to the previous one with the one that uses total_duration:


2023-02-09T12:34:00+01:00
Running bin/kpsr_benchmark
Run on (4 X 2904 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x2)
  L1 Instruction 32 KiB (x2)
  L2 Unified 256 KiB (x2)
  L3 Unified 4096 KiB (x1)
Load Average: 0.66, 0.45, 0.29
----------------------------------------------------------------------------------------------------------
Benchmark                                                                         Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------------------------------
FactoryPerformanceTest/SimpleTest/0/manual_time       98.9 ms         16.4 ms           10 listenerEvents=9 publishedEvents=10
FactoryPerformanceTest/SimpleTest/1/manual_time        386 ms         15.7 ms           10 listenerEvents=9 publishedEvents=10

First case corresponds to the previous version. Now total time is bigger than before, correct me if I'm wrong but I think it may be owing to the fact that I was setting the accumulated duration instead of the one that belongs to each iteration. There's also the fact that the listener (in the previous case where set is done after publish is) is running in another thread and could have being setting the time after state.setIterationTime was done